cli/src/cmd/snapshot.rs

```diff
@@ -0,0 +1,274 @@
+//! Snapshot command
+
+use crate::cmd::{
+    test,
+    test::{Test, TestOutcome},
+    Cmd,
+};
+use once_cell::sync::Lazy;
+use regex::Regex;
+use std::{
+    collections::HashMap,
+    fmt::Write,
+    fs,
+    io::{self, BufRead},
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+use structopt::StructOpt;
+
+/// A regex that matches a basic snapshot entry like
+/// `testDeposit() (gas: 58804)`
+pub static RE_BASIC_SNAPSHOT_ENTRY: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"(?P<sig>(\w+)\s*\((.*?)\))\s*\((gas:)?\s*(?P<gas>\d+)\)").unwrap());
+
+#[derive(Debug, Clone, StructOpt)]
+pub struct SnapshotArgs {
+    /// All test arguments are supported
+    #[structopt(flatten)]
+    test: test::TestArgs,
+    /// Additional configs for test results
+    #[structopt(flatten)]
+    config: SnapshotConfig,
+    #[structopt(
+        help = "Compare against a snapshot and display changes from the snapshot. Takes an optional snapshot file, [default: .gas-snapshot]",
+        long
+    )]
+    diff: Option<Option<PathBuf>>,
+    #[structopt(help = "How to format the output.", long)]
+    format: Option<Format>,
+    #[structopt(help = "Output file for the snapshot.", default_value = ".gas-snapshot", long)]
+    snap: PathBuf,
+}
+
+impl Cmd for SnapshotArgs {
+    type Output = ();
+
+    fn run(self) -> eyre::Result<()> {
+        let outcome = self.test.run()?;
+        outcome.ensure_ok()?;
+        let tests = self.config.apply(outcome);
+
+        match self.diff {
+            Some(Some(snap)) => {
+                let snaps = read_snapshot(snap)?;
+                diff(tests, snaps)?;
+            }
+            Some(None) => {
+                let snaps = read_snapshot(self.snap)?;
+                diff(tests, snaps)?;
+            }
+            _ => {
+                write_to_snapshot_file(&tests, self.snap)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+// TODO implement pretty tables
+#[derive(Debug, Clone)]
+pub enum Format {
+    Table,
+}
+
+impl FromStr for Format {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "t" | "table" => Ok(Format::Table),
+            _ => Err(format!("Unrecognized format `{}`", s)),
+        }
+    }
+}
+
+#[derive(Debug, Clone, StructOpt)]
+pub struct CheckSnapshotArgs {
+    #[structopt(
+        help = " Input gas snapshot file to compare against.",
+        default_value = ".gas_snapshot",
+        short,
+        long
+    )]
+    input: PathBuf,
+    #[structopt(flatten)]
+    config: SnapshotConfig,
+}
+
+impl Cmd for CheckSnapshotArgs {
+    type Output = ();
+
+    fn run(self) -> eyre::Result<()> {
+        dbg!(self);
+        todo!()
```

Comment:
What is the expected behavior for this?

Reply:
intended for `check-snapshot` as mentioned here https://github.com/dapphub/dapptools/issues/761
> Adding a dapp check-snapshot command that runs dapp snapshot (but outputs to a temp file instead of .dappshots) and compares it with .dappshots. If they don't match the command exits with an error.